File: backend/app/schemas/pre_feed_cost.py
from datetime import date, datetime
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field

from app.schemas.common import ConfidenceLevel, DataStatus
from app.schemas.document import DocumentRead


CostType = Literal["capex", "opex"]
CostComponent = Literal[
    "capture_package",
    "electrolyzer",
    "methanol_plant",
    "storage_port",
    "grid_power",
    "land",
    "mrv",
    "other",
]
OpexRecurrence = Literal["annual", "monthly", "quarterly", "weekly", "daily", "one_time"]
SelectionType = Literal["blended_package", "vendor_proposal"]

COST_TYPE_VALUES = ("capex", "opex")
COST_COMPONENT_VALUES = (
    "capture_package",
    "electrolyzer",
    "methanol_plant",
    "storage_port",
    "grid_power",
    "land",
    "mrv",
    "other",
)
OPEX_CATEGORY_VALUES = (
    "fixed_opex",
    "variable_opex",
    "electricity",
    "water",
    "chemicals",
    "labor",
    "maintenance",
    "transport",
    "mrv",
    "other",
)
RECURRENCE_VALUES = ("annual", "monthly", "quarterly", "weekly", "daily", "one_time")
SELECTION_TYPE_VALUES = ("blended_package", "vendor_proposal")


class PreFeedCostItemCreate(BaseModel):
    vendor_proposal_id: str | None = None
    cost_type: CostType
    cost_component: str = Field(min_length=1, max_length=80)
    amount: float = Field(ge=0)
    currency: str = Field(default="USD", min_length=3, max_length=8)
    unit_basis: str | None = Field(default=None, max_length=80)
    recurrence: OpexRecurrence | None = None
    contingency_percent: float | None = Field(default=None, ge=0)
    escalation_percent: float | None = Field(default=None, ge=0)
    source_label: str | None = Field(default=None, max_length=160)
    data_status: DataStatus = "unknown"
    confidence_level: ConfidenceLevel = "unknown"
    notes: str | None = None


class PreFeedCostItemUpdate(BaseModel):
    vendor_proposal_id: str | None = None
    cost_type: CostType | None = None
    cost_component: str | None = Field(default=None, min_length=1, max_length=80)
    amount: float | None = Field(default=None, ge=0)
    currency: str | None = Field(default=None, min_length=3, max_length=8)
    unit_basis: str | None = Field(default=None, max_length=80)
    recurrence: OpexRecurrence | None = None
    contingency_percent: float | None = Field(default=None, ge=0)
    escalation_percent: float | None = Field(default=None, ge=0)
    source_label: str | None = Field(default=None, max_length=160)
    data_status: DataStatus | None = None
    confidence_level: ConfidenceLevel | None = None
    notes: str | None = None


class PreFeedCostItemRead(BaseModel):
    id: str
    package_id: str
    vendor_proposal_id: str | None
    cost_type: CostType
    cost_component: str
    amount: float
    currency: str
    unit_basis: str | None
    recurrence: OpexRecurrence | None
    contingency_percent: float | None
    escalation_percent: float | None
    source_label: str | None
    data_status: DataStatus
    confidence_level: ConfidenceLevel
    notes: str | None
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class PreFeedCostSummaryRead(BaseModel):
    package_id: str
    vendor_proposal_id: str | None = None
    item_count: int
    capex_total_by_currency: dict[str, float]
    annual_opex_total_by_currency: dict[str, float]
    scenario_ready_assumptions: dict[str, float]
    warnings: list[str]


class PreFeedVendorProposalCreate(BaseModel):
    supporting_document_id: str | None = None
    vendor_name: str = Field(min_length=1, max_length=160)
    proposal_name: str = Field(min_length=1, max_length=160)
    scope_capture_package: bool = False
    scope_electrolyzer: bool = False
    scope_methanol_plant: bool = False
    scope_storage_port: bool = False
    scope_grid_power: bool = False
    scope_land: bool = False
    scope_mrv: bool = False
    commercial_basis: str | None = None
    delivery_assumptions: str | None = None
    exclusions: str | None = None
    validity_date: date | None = None
    data_status: DataStatus = "unknown"
    confidence_level: ConfidenceLevel = "unknown"
    notes: str | None = None


class PreFeedVendorProposalUpdate(BaseModel):
    supporting_document_id: str | None = None
    vendor_name: str | None = Field(default=None, min_length=1, max_length=160)
    proposal_name: str | None = Field(default=None, min_length=1, max_length=160)
    scope_capture_package: bool | None = None
    scope_electrolyzer: bool | None = None
    scope_methanol_plant: bool | None = None
    scope_storage_port: bool | None = None
    scope_grid_power: bool | None = None
    scope_land: bool | None = None
    scope_mrv: bool | None = None
    commercial_basis: str | None = None
    delivery_assumptions: str | None = None
    exclusions: str | None = None
    validity_date: date | None = None
    data_status: DataStatus | None = None
    confidence_level: ConfidenceLevel | None = None
    notes: str | None = None


class PreFeedVendorProposalRead(BaseModel):
    id: str
    package_id: str
    supporting_document_id: str | None
    vendor_name: str
    proposal_name: str
    scope_capture_package: bool
    scope_electrolyzer: bool
    scope_methanol_plant: bool
    scope_storage_port: bool
    scope_grid_power: bool
    scope_land: bool
    scope_mrv: bool
    commercial_basis: str | None
    delivery_assumptions: str | None
    exclusions: str | None
    validity_date: date | None
    data_status: DataStatus
    confidence_level: ConfidenceLevel
    notes: str | None
    supporting_document: DocumentRead | None = None
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class PreFeedVendorGapRead(BaseModel):
    proposal_id: str
    source_module: str
    missing_data_name: str
    impact_level: str
    priority_level: str
    recommendation: str
    status: str
    confidence_level: ConfidenceLevel


class PreFeedVendorComparisonRead(BaseModel):
    proposal_id: str
    vendor_name: str
    proposal_name: str
    capex_total_by_currency: dict[str, float]
    annual_opex_total_by_currency: dict[str, float]
    scope_completeness_score: float
    missing_scopes: list[str]
    gap_count: int
    confidence_level: ConfidenceLevel
    validity_date: date | None


class PreFeedCostBasisSelectionCreate(BaseModel):
    package_id: str
    vendor_proposal_id: str | None = None
    selection_type: SelectionType
    selected_by: str | None = Field(default=None, max_length=160)
    selection_notes: str | None = None


class PreFeedCostBasisSelectionRead(BaseModel):
    id: str
    scenario_id: str
    package_id: str
    vendor_proposal_id: str | None
    selection_type: SelectionType
    is_active: bool
    snapshot_totals: dict[str, Any] | None
    scenario_ready_assumptions: dict[str, Any] | None
    selected_by: str | None
    selection_notes: str | None
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)

File: frontend/types/prefeed-cost.ts
import type { ProjectDocument } from "@/types/document";
import type { ConfidenceLevel, DataStatus } from "@/types/plant";

export type PreFeedCostType = "capex" | "opex";
export type PreFeedOpexRecurrence = "annual" | "monthly" | "quarterly" | "weekly" | "daily" | "one_time";
export type PreFeedSelectionType = "blended_package" | "vendor_proposal";

export type PreFeedCostItem = {
  id: string;
  package_id: string;
  vendor_proposal_id: string | null;
  cost_type: PreFeedCostType;
  cost_component: string;
  amount: number;
  currency: string;
  unit_basis: string | null;
  recurrence: PreFeedOpexRecurrence | null;
  contingency_percent: number | null;
  escalation_percent: number | null;
  source_label: string | null;
  data_status: DataStatus;
  confidence_level: ConfidenceLevel;
  notes: string | null;
  created_at: string;
  updated_at: string;
};

export type PreFeedCostItemPayload = {
  vendor_proposal_id?: string | null;
  cost_type: PreFeedCostType;
  cost_component: string;
  amount: number;
  currency: string;
  unit_basis?: string | null;
  recurrence?: PreFeedOpexRecurrence | null;
  contingency_percent?: number | null;
  escalation_percent?: number | null;
  source_label?: string | null;
  data_status: DataStatus;
  confidence_level: ConfidenceLevel;
  notes?: string | null;
};

export type PreFeedCostSummary = {
  package_id: string;
  vendor_proposal_id: string | null;
  item_count: number;
  capex_total_by_currency: Record<string, number>;
  annual_opex_total_by_currency: Record<string, number>;
  scenario_ready_assumptions: Record<string, number>;
  warnings: string[];
};

export type PreFeedVendorProposal = {
  id: string;
  package_id: string;
  supporting_document_id: string | null;
  vendor_name: string;
  proposal_name: string;
  scope_capture_package: boolean;
  scope_electrolyzer: boolean;
  scope_methanol_plant: boolean;
  scope_storage_port: boolean;
  scope_grid_power: boolean;
  scope_land: boolean;
  scope_mrv: boolean;
  commercial_basis: string | null;
  delivery_assumptions: string | null;
  exclusions: string | null;
  validity_date: string | null;
  data_status: DataStatus;
  confidence_level: ConfidenceLevel;
  notes: string | null;
  supporting_document: ProjectDocument | null;
  created_at: string;
  updated_at: string;
};

export type PreFeedVendorProposalPayload = {
  supporting_document_id?: string | null;
  vendor_name: string;
  proposal_name: string;
  scope_capture_package: boolean;
  scope_electrolyzer: boolean;
  scope_methanol_plant: boolean;
  scope_storage_port: boolean;
  scope_grid_power: boolean;
  scope_land: boolean;
  scope_mrv: boolean;
  commercial_basis?: string | null;
  delivery_assumptions?: string | null;
  exclusions?: string | null;
  validity_date?: string | null;
  data_status: DataStatus;
  confidence_level: ConfidenceLevel;
  notes?: string | null;
};

export type PreFeedVendorGap = {
  proposal_id: string;
  source_module: string;
  missing_data_name: string;
  impact_level: string;
  priority_level: string;
  recommendation: string;
  status: string;
  confidence_level: ConfidenceLevel;
};

export type PreFeedVendorComparison = {
  proposal_id: string;
  vendor_name: string;
  proposal_name: string;
  capex_total_by_currency: Record<string, number>;
  annual_opex_total_by_currency: Record<string, number>;
  scope_completeness_score: number;
  missing_scopes: string[];
  gap_count: number;
  confidence_level: ConfidenceLevel;
  validity_date: string | null;
};

export type PreFeedActiveCostBasis = {
  id: string;
  scenario_id: string;
  package_id: string;
  vendor_proposal_id: string | null;
  selection_type: PreFeedSelectionType;
  is_active: boolean;
  snapshot_totals: Record<string, unknown> | null;
  scenario_ready_assumptions: Record<string, unknown> | null;
  selected_by: string | null;
  selection_notes: string | null;
  created_at: string;
  updated_at: string;
};

export type PreFeedActiveCostBasisPayload = {
  package_id: string;
  vendor_proposal_id?: string | null;
  selection_type: PreFeedSelectionType;
  selected_by?: string | null;
  selection_notes?: string | null;
};
